### PR TITLE
General clean up, bug fixes.

### PR DESCRIPTION
**Pull request by gdborton**
Changeset sort of expanded as I discovered more stuff while working... Most of the commits have fairly descriptive messages.  Let me know if you want me to break this up, or copy/pasta the commits logs into the PR.

### CODE REVIEW

**Comment by gdborton** [comment]
This commit might end up breaking this repo in the future assuming that it gets fixed upstream in Jest.

https://github.com/ljharb/jest-wrap/pull/4/commits/e71496de6db9eb22ed11bd6d6870f6a68c19fdbf

Personally I think it's better to fix it now, and then fix it again later, but I'm not sure how you'd handle that w/ semver.

**Comment by ljharb** [comment]
I pulled in the specify fix with a tweak and rebased the PR.

File: index.js [hunk 1]
```diff
@@ -133,11 +133,13 @@ var createAssertion = function createAssertion(type, message, wrappers, block, m
 	if (descriptors.length === 0) {
 		throw new RangeError(inspect(type) + ' called with no wrappers defined');
 	}
-	var describeMsgs = [];
-	forEach(wrappers, function (wrapper) {
-		checkThis(wrapper);
-		describeMsgs.push(getThisDescription(wrapper));
-	});
+
+	var describeMsgs = descriptors.reduce(function descriptorReducer(descriptions, descriptor) {
```
Comment (ljharb) [inline index.js]: Using `.reduce` means it requires ES5; my hope is that jest will eventually transpile and work on older engines. I'd prefer to keep using `forEach` here.

File: index.js [hunk 2]
```diff
@@ -133,11 +133,13 @@ var createAssertion = function createAssertion(type, message, wrappers, block, m
 	if (descriptors.length === 0) {
 		throw new RangeError(inspect(type) + ' called with no wrappers defined');
 	}
-	var describeMsgs = [];
-	forEach(wrappers, function (wrapper) {
-		checkThis(wrapper);
-		describeMsgs.push(getThisDescription(wrapper));
```
Comment (ljharb) [inline index.js]: is the only change here that it's not pushing `getThisDescription(wrapper)` when the description is falsy? If so, I'd prefer a more minimal change here.

Comment (gdborton) [inline index.js]: I changed extend to apply the description to the descriptor object in all cases, so we don't need to look at wrappers anymore, just at the flattened `descriptors`.

Comment (ljharb) [inline index.js]: hmm - this means it's mutating the descriptor object the user passes in; and I'm concerned that the uses of the `concatThis` helper (in `skip` and `only`) without `extend` will mean that those places aren't persisting the description.

Could we fix this without mutating `descriptor` or changing how descriptions are stored?

Comment (gdborton) [inline index.js]: `concatThis` copies the wrappers array, which is where the descriptors are stored so they should get copied over.

We're also already mutating the descriptor object by turning each of its key/values into arrays. https://github.com/ljharb/jest-wrap/blob/master/index.js#L220

We could probably shrink the api here a bit by changing `.extend` to just take a descriptor, and .`use` to take a function that returns a descriptor, then validating that descriptor.

**Comment by gdborton** [comment]
@ljharb Updated, looking for some feedback on https://github.com/ljharb/jest-wrap/pull/4#discussion_r110650301

I'd like to see this merged soon, so lemme know if/how you want me to rework.

File: index.js (hunk 2 above)
Comment (ljharb) [inline index.js]: Gross, you're right. I'm not a good programmer.

Hmm - then the common case of `this.use(wrapperA, arg1, arg2).use(wrapperB, arg1, arg2).extend('description')` becomes `this.use(wrapperA, arg1, arg2).use(wrapperB, arg1, arg2).extend({ description: 'description' })`, which is slightly more verbose. I'm not sure what you mean about `.use` taking a descriptor; `use` is meant to take an existing plugin, not a descriptor.

OK, so - let's keep your change that modifies `descriptor`, for sure. However, I think that means we should remove the `getThisDescription` and `setThisDescription` helpers entirely?

Comment (gdborton) [inline index.js]: `.use` should take a function that returns a descriptor, not take a descriptor directly.  That will prevent having any subwrappers.

Comment (gdborton) [inline index.js]: Updated PR to remove description helpers, tests still pass! 😌

**Pull request merged**